### DISCUSSION TITLE
Content collection API for adding entries programmatically

### DISCUSSION THREAD

**Discussion by delucis** [withastro/roadmap#688] [Proposal]
### Body

# Summary

Content collections currently support on-disk files. This proposal suggests an API for adding collection entries programmatically.

# Background & Motivation

Content collections offer a powerful way to use content in Astro and associate strongly typed metadata with each piece of content via Markdown frontmatter. This makes it easy to build tools that can flexibly render user content based on its structured data. Currently, this is limited to content that lives on disk in the `src/content/` directory of a repo.

As we’ve built Starlight we’ve seen users like the general experience of working this way, but several have expressed a desire to add pages from additional sources to Starlight sites. For example:
- pages that are [auto-generated from source code using a tool like TypeDoc](https://github.com/HiDeoo/starlight-typedoc)
- pages that are located elsewhere in a (mono)repo ([workaround using symlinks](https://www.eliostruyf.com/symlink-content-astro-portability/))
- pages that are loaded from a remote API, e.g headless CMS
- pages that are programmatically generated from an alternative data format (e.g. on-disk JSON)

Historically, we’ve solved many of these cases with dynamic routes in a project or `injectRoute` for an integration. However, for a project like Starlight where frontmatter is used to build site navigation, make layout decisions, etc. those tools place content “out of sight” for the framework. An API for adding entries to a content collection programmatically would support something similar to `injectRoute` but specifically for structured content and metadata.

# Goals

- Support adding content to content collections from other sources such that code consuming a collection can treat it the same way as it would content from files in the repo.

# Example

There are lot of fairly complex technical considerations around how to implement this, but I’m currently imagining an integration hook similar to `injectRoute()` like `injectContent()`. There may be a question of whether the code should run on dev server start or be evaluated lazily when a collection is used. Showing a callback style here for the latter case.

```js
const integration = {
  name: 'simple-programmatic-content',
  hooks: {
    'astro:config:setup': ({ injectContent }) => {
      injectContent('docs', () => {
        return [
          {
            id: 'injected.md',
            slug: 'injected',
            data: { title: 'Injected page' },
            content: 'This page was **injected** via an integration!', 
          },
        ];
      });
    },
  },
};
```

Not going to go into too much more detail for now, but imagine the callback could be `async` for file system or network operations.

During build the callback would need to be evaluated while building static endpoints and the result bundled for SSR, I guess? Don’t know if that’s an acceptable design, but it aligns with the static-build behaviour of content collections currently. Very much open to other suggestions.

This could also take a form closer to `injectRoute` where instead of an inline callback, the integration provides a path to a module that contains the logic for loading content entries, i.e. `injectContent('docs', './get-docs.ts')`.

**Comment by maddsua** [comment]
I can totally see this being useful

**Comment by lilnasy** [comment]
I really like this idea!

It could benefit one of the most common use cases for astro - rendering from a headless CMS. If an integration can take control of making the API calls, it can also be smart about caching them - potentially saving several minutes from each build.

**Comment by delucis** [comment]
Adding a note from chatting with @matthewp earlier today:

One challenge could be HMR: during dev an integration might need to use `addWatchFile()` for local file-based generation; what would that look like for remote content?

**Comment by chrisrueger** [comment]
I originally posted here https://github.com/withastro/starlight/discussions/962#discussioncomment-7659348 and found this discussion based on the responses I got. So I try to add my feedback here too, to outline usecases.

> It could benefit one of the most common use cases for astro - rendering from a headless CMS.

Yes.
I want to create documentation site basedon data from an external API.... similar to https://github.com/HiDeoo/starlight-openapi 

My first step was just doing plain astro pages and dynamic routes. This technically works but I struggle to get that into the Starlight layout. 

- generate documentation based on an external API call (e.g. `fetch(https://example.com/api/items)` returning JSON
- the result could extend the sidebar, generate a list-page (e.g. `myitems.astro`) and related item-details pages (e.g. `[myitemslug].astro`)

Additionally I'd like to add a special "hybrid API-document-merge approach":

- for some items I would like to **enrich** the API-generated pages which additional prose-content from `/docs/enrich/myitemslug.md` pages which gets merged with the details pages by slug coming from the API. 

I already solved this hybrid-merge-approach with the full astrojs power (`getCollection('docs').getEntry('...')`) in a dynamic route page but I currently do not get it into the starlight layout. 

> There may be a question of whether the code should run on dev server start or be evaluated lazily when a collection is used

Once at startup would be great, since in our use case we fetch all the data we need in one large batch.

**Reply by chrisrueger** [reply]
Just one additional thought: I used Gridsome in the past and they are solving this quite good in my opinion. See https://gridsome.org/docs/fetching-data/

In Gridsome a Collection can be created from different sources (file system, .md files,XML, external API etc.). Then this collection can be queried to build page content. 
Maybe helpful to get ideas.

**Comment by DarrenBaldwin07** [comment]
I like this idea a lot!

**Comment by delucis** [comment]
I think most of the ideas here will likely end up being covered by #946 so going to close this one in favour of that!
